packages/wonder-blocks-data/src/hooks/use-shared-cache.js

```diff
@@ -0,0 +1,106 @@
+// @flow
+import * as React from "react";
+import {KindError, Errors} from "@khanacademy/wonder-stuff-core";
+import {InMemoryCache} from "../util/in-memory-cache.js";
+import type {ValidData} from "../util/types.js";
+
+/**
+ * A function for inserting a value into the cache or clearing it.
+ */
+type CacheValueFn<TValue: ValidData> = (value: ?TValue) => void;
+
+/**
+ * This is the cache.
+ * It's incredibly complex.
+ * Very in-memory. So cache. Such complex. Wow.
+ */
+const cache = new InMemoryCache();
+
+/**
+ * Clear the in-memory cache or a single scope within it.
+ */
+export const clearSharedCache = (scope: string = "") => {
+    // If we have a valid scope (empty string is falsy), then clear that scope.
+    if (scope && typeof scope === "string") {
+        cache.purgeScope(scope);
+    } else {
+        // Just reset the object. This should be sufficient.
+        cache.purgeAll();
+    }
+};
+
+/**
+ * Hook to retrieve data from and store data in an in-memory cache.
+ *
+ * @returns {[?ReadOnlyCacheValue, CacheValueFn]}
+ * Returns an array containing the current cache entry (or undefined), a
+ * function to set the cache entry (passing null or undefined to this function
+ * will delete the entry).
+ *
+ * To clear a single scope within the cache or the entire cache,
+ * the `clearScopedCache` export is available.
+ *
+ * NOTE: Unlike useState or useReducer, we don't automatically update folks
+ * if the value they reference changes. We might add it later (if we need to),
+ * but the likelihood here is that things won't be changing in this cache in a
+ * way where we would need that. If we do (and likely only in specific
+ * circumstances), we should consider adding a simple boolean useState that can
+ * be toggled to cause a rerender whenever the referenced cached data changes
+ * so that callers can re-render on cache changes. However, we should make
+ * sure this toggling is optional - or we could use a callback argument, to
+ * achieve this on an as-needed basis.
+ */
+export const useSharedCache = <TValue: ValidData>(
+    id: string,
+    scope: string,
+    initialValue?: ?TValue | (() => ?TValue),
+): [?TValue, CacheValueFn<TValue>] => {
```

Comment:
Signature matches the return of `useState`, with a result and a setting function. I like this pattern since it allows folks to easily name the function and value something that makes sense for their context. Returning an object allows for that, but it's more cumbersome.

Reply:
Love it! 🙂